File: chapter2/test_arith.ts
import { assert, assertEquals } from "https://deno.land/std@0.224.0/assert/mod.ts";
import { parseArith } from "npm:tiny-ts-parser";
import { typecheck, Type } from "./arith.ts";

const expectResult = (expected: Type) => {
  return (context: Deno.TestContext) => {
    const result = typecheck(parseArith(context.name));
    assertEquals(result, expected);
  }
}

const expectThrow = (expected: string) => {
  return (context: Deno.TestContext) => {
    try {
      typecheck(parseArith(context.name));
    } catch (e) {
      assertEquals(e, expected);
      return;
    }
    assert(false, `Expected error: ${expected}`);
  }
}

Deno.test("1 + 2", expectResult({ tag: "Number" }));
Deno.test("1 + true", expectThrow("number expected"));
Deno.test("1 + (2 + 3)", expectResult({ tag: "Number" }));
Deno.test("1 ? 2 : 3", expectThrow("boolean expected"));
Deno.test("true ? 1 : true", expectThrow("then and else have different types"));

File: lib/chapter2/test_arith.rb
require_relative "./arith"

require "minitest"
include Minitest::Assertions

class << self
  attr_accessor :assertions
end
self.assertions = 0

# TSのテストコードをそのまま貼り付けられるようにする
class Deno
  def self.test(source, proc)
    proc.call(source)
  end
end

def expectResult(expected)
  Proc.new do |source|
    assert_equal(expected, Checker.typecheck(TinyRbParser.parse(source)))
  end
end

def expectThrow(expected_message)
  Proc.new do |source|
    err = assert_raises(RuntimeError) do
      Checker.typecheck(TinyRbParser.parse(source))
    end
    assert_equal(expected_message, err.message)
  end
end

Deno.test("1 + 2", expectResult({ tag: "Number" }));
Deno.test("1 + true", expectThrow("number expected"));
Deno.test("1 + (2 + 3)", expectResult({ tag: "Number" }));
Deno.test("true ? 1 : 2", expectResult({ tag: "Number" }));
Deno.test("true ? false : true", expectResult({ tag: "Boolean" }));
Deno.test("1 ? 2 : 3", expectThrow("boolean expected"));
Deno.test("true ? 1 : true", expectThrow("then and else have different types"));
Deno.test("true ? (1 + 2) : (3 + (false ? 4 : 5))", expectResult({ tag: "Number" }));
